contracts/bonding/BondingCheckpoints.sol

```diff
@@ -0,0 +1,393 @@
+     */
+    function hasCheckpoint(address _account) external view returns (bool) {
+        return bondingCheckpoints[_account].startRounds.length > 0;
+    }
+
+    /**
+     * @notice Called by the BondingManager when the total active stake changes.
+     * @dev This is called only from the {BondingManager-setCurrentRoundTotalActiveStake} function to set the total
+     * active stake in the current round.
+     * @param _totalStake From {BondingManager-currentRoundTotalActiveStake}
+     * @param _round The round for which the total active stake is valid. This is normally the current round.
+     */
+    function checkpointTotalActiveStake(uint256 _totalStake, uint256 _round) public virtual onlyBondingManager {
+        if (_round > clock()) {
+            revert FutureCheckpoint(_round, clock());
+        }
+
+        totalActiveStakeCheckpoints[_round] = _totalStake;
+
+        totalStakeCheckpointRounds.pushSorted(_round);
+    }
+
+    // Historical stake access functions
+
+    /**
+     * @dev Gets the checkpointed total active stake at a given round.
+     * @param _round The round for which we want to get the total active stake.
+     */
+    function getTotalActiveStakeAt(uint256 _round) public view virtual returns (uint256) {
+        if (_round > clock()) {
+            revert FutureLookup(_round, clock());
+        }
+
+        uint256 activeStake = totalActiveStakeCheckpoints[_round];
+
+        if (activeStake == 0) {
+            uint256 lastInitialized = checkedFindLowerBound(totalStakeCheckpointRounds, _round);
+
+            // Check that the round was in fact initialized so we don't return a 0 value accidentally.
+            if (lastInitialized != _round) {
+                revert MissingRoundCheckpoint(_round);
+            }
+        }
+
+        return activeStake;
+    }
+
+    /**
+     * @notice Gets the bonding state of an account at a given round.
+     * @dev In the case of delegators it is the amount they are delegating to a transcoder, while for transcoders this
+     * includes all the stake that has been delegated to them (including self-delegated).
+     * @param _account The account whose bonding state we want to get.
+     * @param _round The round for which we want to get the bonding state. Normally a proposal's vote start round.
+     * @return amount The active stake of the account at the given round including any accrued rewards. In case of
+     * transcoders this also includes all the amount delegated towards them by other delegators.
+     * @return delegateAddress The address the account delegated to. Will be equal to _account in case of transcoders.
+     */
+    function getBondingStateAt(address _account, uint256 _round)
+        public
+        view
+        virtual
+        returns (uint256 amount, address delegateAddress)
+    {
+        BondingCheckpoint storage bond = getBondingCheckpointAt(_account, _round);
```

Comment:
I think this can be `memory` vs `storage` to save gas costs

Reply:
```suggestion
        BondingCheckpoint memory bond = getBondingCheckpointAt(_account, _round);
```

Reply:
Same across the board for a lot of the view functions

Reply:
I think using `memory` here would actually increase gas costs because `getBondingCheckpointAt()` currently returns a storage pointer to a struct which I believe defers gas cost consumption for storage reads until the individual fields of the struct are read. If we use `memory` here I believe the values of the storage fields of the struct would need to all be read and written into memory.

In functions where we need to read the values of the struct fields infrequently I think it makes sense to continue using a storage pointer because the gas overhead from reading/copying all struct fields from storage to memory likely outweighs the gas cost of reading the field values direct from storage. In functions where we need to read the values of the struct fields super frequently then it might make sense to read/copy all struct fields from storage to memory because the gas overhead could be small relative to the savings from repeatedly reading from memory instead of storage.

Reply:
As a sanity check, I ran the gas reports with [the `memory` version](https://github.com/livepeer/protocol/pull/611/commits/ec801e26101678899706ab805760daaf23165173) for these variables and return values, and the cost of `checkpointBondingState` doesn't change much as expected, while the `getBondingStateAt` function increases gas cost in up to 10k (or 7k on average for delegator/transcoder/nonparticipant). So it is in fact more efficient to keep it as a storage so we don't have to copy the whole struct back to memory.

Attaching the report outputs here FTR.
[gas-memory.txt](https://github.com/livepeer/protocol/files/12361511/gas-memory.txt)
[gas-storage.txt](https://github.com/livepeer/protocol/files/12361512/gas-storage.txt)